src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -195,11 +195,15 @@ public static void run(DynamicObject thread, RubyContext context, Node currentNo
             task.run();
         } catch (ThreadExitException e) {
             setThreadValue(context, thread, context.getCoreLibrary().getNil());
-            return;
         } catch (RaiseException e) {
             setException(context, thread, e.getException(), currentNode);
         } catch (ReturnException e) {
             setException(context, thread, context.getCoreExceptions().unexpectedReturn(currentNode), currentNode);
+        } catch (ExitException e) {
+            final Thread rootThread = Layouts.FIBER.getThread(Layouts.THREAD.getFiberManager(context.getThreadManager().getRootThread()).getCurrentFiber());
+            context.getSafepointManager().pauseThreadAndExecute(rootThread, currentNode, (actionThread, actionCurrentNode) -> {
+                throw e;
+            });
```

Comment:
Hum, but this is difficult as that expects a RaiseException.
Please just extract these lines 4 in a method then.
You also need to `setThreadValue(context, thread, context.getCoreLibrary().getNil());` here, otherwise Thread#value would return null.